### PR TITLE
[WIP] 132104697 extract org_owner? method

### PR DESCRIPTION
**Pull request by Maroo-b**
The `org_owner?` method is used for authorization check, so what about using pundit gem to handle the authorization ?

@tansaku currently there is two roles for users: superadmin and simple users ? because I saw another role: admin. Is there a difference between superadmin and admin ?

### CODE REVIEW

**Comment by tansaku** [comment]
@Maroo-b we have this idea of a super-admin having site-wide access, while there may be an organisation admin who just has admin of their organisation - it's the kind of thing that should be exposed in the domain model:

https://github.com/AgileVentures/LocalSupport/blob/develop/docs/domain_vision_statement.md

but ideally we should go through the codebase and talk with the client to make sure this all links up ... I wonder if we can turn the domain model into cucumber stories somehow?  Perhaps we just need to make the domain model more prominent somehow and make sure the domain terminology is used accurately through the cucumber steps

**Comment by Maroo-b** [comment]
@tansaku I'm not familiar with Domain Driven Design I need to check examples of this concept.
I updated the PR to add pundit gem integration for VolunteerOp model, just to show the changes and how it helps to clear controller and model. 
I suggest that I replace old authorization methods with pundit ones. what do you think ?

**Comment by tansaku** [comment]
that sounds good in principle - I need to review - will get to asap

File: app/controllers/invitations_controller.rb [hunk 1]
```diff
@@ -1,8 +1,8 @@
 class InvitationsController < ApplicationController
-  before_filter :authorize
 
   # xhr only, tested in a request spec
   def create
+    authorise :invitation, :create?
```
Comment (johnnymo87) [inline app/controllers/invitations_controller.rb]: Should this be spelled `authorize`, or should all the other ones be spelled `authorise`?

I'm not familiar with pundit. Why does this one need a `:create?` argument while the other ones seem to magically infer the action?

File: app/controllers/volunteer_ops_controller.rb [hunk 2]
```diff
@@ -73,29 +76,6 @@ def restrict_by_feature_scope
     :local_only
   end
 
-  def authorize
-    # set @organisation
-    # then can make condition:
-    # unless current_user.can_edit? organisation
-    unless org_owner? || superadmin?
-      flash[:error] = 'You must be signed in as an organisation owner or ' \
-                      'site superadmin to perform this action!'
```
Comment (johnnymo87) [inline app/controllers/volunteer_ops_controller.rb]: Hmm, it looks like this specific message will be lost. Is it possible for this entity's pundit to raise a more specific error, and have this entity's controller rescue it with its own error message? Actually, would it be possible to define the message passed to `flash[:error]` on these pundit objects? Maybe the controller code read something like this:
```
rescue_from InvitationPolicy::NotAuthorized, with: InvitationPolicy::NotAuthorizedMessage
```

**Comment by johnnymo87** [comment]
At work, we use a handrolled policies library (developed by @cored) that somewhat resembles pundit. One difference I see is reflected by this directory structure:
```
▾ app/
  ▾ lib/
    ▾ policies/
      ▾ time_off/
          can_edit_my_request.rb
          can_edit_request.rb
          can_edit_your_request.rb
          can_request.rb
          can_request_on_behalf_of.rb
          can_respond_to_request.rb
          can_view_buckets_and_usages.rb
          can_view_plan.rb
          can_view_request.rb
          enabled.rb
        policy.rb
        policy_mixin.rb
```
With our policies implementation, each policy is a separate object. Each one has a `#valid?` and a `valid!` method, and by default our domain object call `#valid!`. `#valid!` raises an error, which gets rescued in the controller like this:
```ruby
  rescue_from Policies::TimeOff::Enabled::Denied,
    Policies::TimeOff::CanRequest::Denied,
    Policies::TimeOff::CanRequestOnBehalfOf::Denied,
    Policies::TimeOff::CanViewRequest::Denied,
    Policies::TimeOff::CanEditRequest::Denied,
    Policies::TimeOff::CanRespondToRequest::Denied do |e|
      render_error(:unauthorized, e.message)
  end
```
One thing I like about this style is that it strictly adheres to the single responsibility principle. Each policy object validates only one policy. Also, it is declarative. When reading code, I know exactly what is being validated when I see a `Policies::TimeOff::CanEditRequest` be passed into a domain object.

One reservation I have about this pundit implementation is how a given policy, `VolunteerOpsPolicy` for instance, is potentially responsible for all the validations for the `VolunteerOp` entity. Also, I am seeing duplication between the `#create?`, `#update?`, and `#destroy?` methods, which is possibly hiding something that might be expressed as `Policies::VolunteerOps::CanEdit`.

**Comment by tansaku** [comment]
@Maroo-b for this specific refactoring ticket why don't we just do the simplest thing possible and just move the model to the method - get that merged, and we can open a new ticket for a separate refactoring with a different gem?

**Comment by tansaku** [comment]
BTW, it's always great to have links to the PT ticket in the description: https://www.pivotaltracker.com/n/projects/742821/stories/132104697

File: app/controllers/invitations_controller.rb (hunk 1 above)
Comment (Maroo-b) [inline app/controllers/invitations_controller.rb]: You are right, it's a typo it should be authorize.

Usually we check authorization based on a model but in this case it's based on a controller action that's why I'm using Headless policies https://github.com/elabs/pundit#headless-policies

I tested that Headless policies can infer the name of controller action too.

File: app/controllers/volunteer_ops_controller.rb (hunk 2 above)
Comment (Maroo-b) [inline app/controllers/volunteer_ops_controller.rb]: with pundit I rescue the Error with a similar method in application controller

**Comment by Maroo-b** [comment]
@johnnymo87 that's interesting is that library open sourced ? For Rails apps I'm used to CanCan or Pundit, I saw that Authority gem started to gain some interest https://github.com/nathanl/authority

The drawback of using CanCan is using a single Ability file and hard to extend.
For Pundit it's based on REST action, for example you can have different set of rules based on current_user and record for each action (create, update,..) 

In our case there is a duplication because it's based on the same set of access rules.

@tansaku it sounds good idea, but I estimate to finish the refactoring this evening and I'll open  tickets for other small fixes. Also I noticed there is an authorization bug in the current application where a signed in user can access organisation new form via I should not can access https://www.harrowcn.org.uk/organisations/new

**Comment by johnnymo87** [comment]
@Maroo-b No, it's not open source. But for the most part, it's just a bunch of POROs, so we could always handroll our own here, possibly with an existing gem mixed in.

**Comment by tansaku** [comment]
@Maroo-b even if you complete the refactoring, the larger the set of changes the more work it is for me to assess it - this ticket is specifically about simplifying this one method to reduce code climate complexity right? https://www.pivotaltracker.com/n/projects/742821/stories/132104697  I'd love to get that fixed and then have a vote (and hence broader discussion) on using cancan or pundit gems to manage authorisation

**Comment by Maroo-b** [comment]
@tansaku sure, I'll open a new PR.

**Pull request closed without merging**